ops/model.py

```diff
@@ -2998,6 +3042,44 @@ def secret_remove(self, id: str, *, revision: Optional[int] = None):
             args.extend(['--revision', str(revision)])
         self._run_for_secret('secret-remove', *args)
 
+    def open_port(self, protocol: str, port: Optional[int] = None):
+        arg = f'{port}/{protocol}' if port is not None else protocol
```

Comment:
I'm pretty sure port is only optional if protocol is 'icmp' in which case it can't have a port.
I don't know if there is a clearer way to define this.

This is validated by your logic in "_parse_opened_port" that does have 'icmp' and others hard coded for that behavior (must have a port if tcp or udp, must not have a port for icmp)

Reply:
This is noted in the docstring of open-port and (updated docstring for) close-port. It is checked by Juju's open-port and close-port hooks tools and will return an error there. In the Harness I test it up-front, so if folks get this wrong it'll be raised as an error immediately when the unit tests are run.